transformer_engine/jax/cpp_extensions.py

```diff
@@ -210,2265 +248,3369 @@ def custom_caller(name, args, opaque, has_side_effect, **kwargs):
 
-        i_aval, _ = ctx.avals_in
-        i_type = ir.RankedTensorType(inputs.type)
-        i_shape = i_type.shape
-        assert len(i_shape) == shape_rank
-        batch = i_shape[0]
-        heads = i_shape[1]
-        q_seqlen = i_shape[2]
-        k_seqlen = i_shape[3]
+    @staticmethod
+    def batcher(batched_args, batch_dims):
+        """
+        dgated_gelu batcher
+        """
+        assert DgatedGeluPrimitive.outer_primitive is not None
+        dz, x = batched_args
+        _, x_bdim = batch_dims
 
-        mask_type = ir.RankedTensorType(mask.type)
-        mask_shape = mask_type.shape
-        assert len(mask_shape) == shape_rank
-        pad_batch = mask_shape[0]
+        out_bdims = x_bdim
+        return DgatedGeluPrimitive.outer_primitive.bind(dz, x), out_bdims
 
-        out_types = [ir.RankedTensorType.get(i_shape, i_type.element_type)]
-        operands = [inputs, mask]
-        operand_shapes = [i_shape, mask_shape]
-        args = CustomCallArgsWrapper(out_types, operands, operand_shapes)
+    @staticmethod
+    def infer_sharding_from_operands(mesh, arg_infos, result_infos):
+        """
+        dgated_gelu infer_sharding_from_operands
+        """
+        del result_infos    # Unused.
+        gelu_out_spec = get_padded_spec(arg_infos[1])
+        dx_sharding = NamedSharding(mesh, PartitionSpec(*gelu_out_spec))
+        return dx_sharding
 
-        opaque = transformer_engine_jax.pack_softmax_descriptor(batch, pad_batch, heads, q_seqlen,
-                                                                k_seqlen,
-                                                                jax_dtype_to_te_dtype(i_aval.dtype),
-                                                                scale_factor)
+    @staticmethod
+    def partition(mesh, arg_infos, result_infos):
+        """
+        dgated_gelu partition
+        """
+        del result_infos
+        dx_sharding = NamedSharding(mesh, PartitionSpec(*get_padded_spec(arg_infos[1])))
+        arg_shardings = tuple(arg_i.sharding for arg_i in arg_infos)
+        out_shardings = dx_sharding
+        impl = DgatedGeluPrimitive.impl
+        return mesh, impl, out_shardings, arg_shardings
 
-        out = custom_caller(ScaledMaskedSoftmaxFwdPrimitive.name, args, opaque, False)
 
-        return [out]
+register_primitive(DgatedGeluPrimitive)
 
 
-_scaled_masked_softmax_fwd_p = register_primitive(ScaledMaskedSoftmaxFwdPrimitive)
+def dgated_gelu(inputs: jnp.ndarray, gelu_inputs: jnp.ndarray) -> jnp.ndarray:
+    """
+    dgated_gelu fusion wrapper
+    Return dgeglu(inputs)
+    """
+    return DgatedGeluPrimitive.outer_primitive.bind(inputs, gelu_inputs)
 
 
-def scaled_masked_softmax_fwd(inputs: jnp.ndarray, mask: jnp.ndarray,
-                              scale_factor: float) -> jnp.ndarray:
+def _normalize_axis_boundary(axis, ndim):
+    return axis if axis >= 0 else ndim + axis
+
+
+def _multidim_transpose(shape, static_axis_boundary, transpose_axis_boundary):
```

Comment:
static_axis_boundary, what this is? Can you document it somewhere?

Reply:
Refer to [_multidim_transpose/doc](https://github.com/NVIDIA/TransformerEngine/blob/93e5977af4d0ae7bd09aca078edfe7bbd1a1403f/transformer_engine/jax/cpp_extensions.py#L2614) for details.
Basically, it is used to indicate the dim of pipeline parallelism.
For example, A weight in shape of `(hidden_in, hidden_out)`, then it would be `(pipeline_stage, hidden_in, hidden_out)` when pipeline parallelism enabled, and its transpose should be `(pipeline_stage, hidden_out, hidden_in)`. Then the `static_axis_boundary = 0` and `transpose_axis_boundary = 2`